File: backend/app/balance.py
import json
from pathlib import Path
LEGACY=json.loads((Path(__file__).resolve().parents[2]/'content/balance-v1.json').read_text(encoding='utf-8'))

def migrate(snapshot, config):
    if snapshot.balanceRevision>=config.get('balanceRevision',1):
        return
    for b in snapshot.buildings:
        old=LEGACY['buildings'].get(b.type); new=config['buildings'].get(b.type)
        if not old or not new: continue
        old_hp=old['branches'][b.branch].get('hp',old['hp']) if b.branch>=0 else old['hp']
        new_hp=new['branches'][b.branch].get('hp',new['hp']) if b.branch>=0 else new['hp']
        b.hp*=new_hp/old_hp
    for e in snapshot.enemies:
        old=LEGACY['enemies'].get(e.type); new=config['enemies'].get(e.type)
        if not old or not new: continue
        ratio=new['hp']/old['hp']
        e.hp*=ratio; e.maxHp*=ratio
        e.shield*=config['schedule']['reflectShield']/LEGACY['schedule']['reflectShield'] if e.type=='reflector' else ratio
        e.armor=min(.6,max(0,e.armor-old['armor']+new['armor']))
        e.reward=round(e.reward*new['reward']/old['reward'])
    snapshot.balanceRevision=config.get('balanceRevision',1)

File: frontend/src/game/balance.ts
import legacy from '../../../content/balance-v1.json' with {type:'json'};
import type {Config,World} from './types';
export function migrateBalance(s:World,c:Config){
 if((s.balanceRevision??1)>=(c.balanceRevision??1))return;
 for(const b of s.buildings){const old=legacy.buildings[b.type as keyof typeof legacy.buildings],next=c.buildings[b.type];if(!old||!next)continue;const a={...old,...(b.branch>=0?old.branches[b.branch]:{})},z={...next,...(b.branch>=0?next.branches[b.branch]:{})};b.hp*=z.hp/a.hp;}
 for(const e of s.enemies){const old=legacy.enemies[e.type as keyof typeof legacy.enemies],next=c.enemies[e.type];if(!old||!next)continue;const ratio=next.hp/old.hp;e.hp*=ratio;e.maxHp*=ratio;e.shield*=e.type==='reflector'?c.schedule.reflectShield/legacy.schedule.reflectShield:ratio;e.armor=Math.min(.6,Math.max(0,e.armor-old.armor+next.armor));e.reward=Math.round(e.reward*next.reward/old.reward);}
 s.balanceRevision=c.balanceRevision??1;
}
